### DISCUSSION TITLE
Can Yocto allows to upgrade the kernel and / or the CUDA libraries of the Jetson Nano ?

### DISCUSSION THREAD

**Discussion by Marietto2008** [OE4T/meta-tegra#1608] [General]
Hello.

my name is Mario (on the nvidia developer forum my nickname is marietto2008) ;

I've worked hard on the Jetson Nano in recent years as a tester and researcher and experimenter. At some point I've been able to upgrade ubuntu 18 to 20 and 20 to 22.04,but I haven't been able to upgrade the kernel of the Jetson nano from 4.9 to 5 and most of all,I haven't been able to upgrade the CUDA libraries from 10 to 11 or 12. I wished to try some new approaches,but then my nano broke. Recently I've got a new Jetson nano to finish the "job" that I'd left behind. And it happened that someone informed me about Yocto. The installation of yocto is a good experiment to try. I want to ask you if it allows you to upgrade the kernel and / or the CUDA libraries.

Thanks.

**Comment by madisongh** [comment]
Sorry, Mario, but unfortunately there's no magic here that would let you run a kernel, or CUDA support, that NVIDIA doesn't actually provide in JetPack.

It might be technically possible to port the NVIDIA additions/modifications from their 4.9-based sources to some newer kernel, but at a huge cost, and no guarantee of success. Plus you'd still be stuck with the older CUDA.

**Comment by Marietto2008** [comment]
Matt,can the following CUDA packages be installed on the Jetson Nano with some tweaking ?

1) https://developer.nvidia.com/cuda-12-0-0-download-archive?target_os=Linux&target_arch=aarch64-jetson&Compilation=Native&Distribution=Ubuntu&target_version=20.04&target_type=deb_local

or :

2) https://developer.nvidia.com/cuda-11-8-0-download-archive?target_os=Linux&target_arch=aarch64-jetson&Compilation=Native&Distribution=Ubuntu&target_version=20.04&target_type=deb_local

(target arm64-sbsa or aarch64-jetson)

or :

3) https://www.seeedstudio.com/blog/2020/07/29/install-cuda-11-on-jetson-nano-and-xavier-nx/

**Reply by madisongh** [reply]
It's possible that they could be installed, and maybe even run.  It wasn't until JetPack 5, though, that NVIDIA supported installing other versions of CUDA (alongside the normal version included in JetPack), and I expect that the 11.8.0 and 12.0.0 downloads are meant for use with that JetPack 5 feature. 

That blog post doesn't have much detail about whether it would actually work or not. And I would expect trouble if you tried to also use TensorRT or any of the other JetPack packages that depend on CUDA.

**Reply by Marietto2008** [reply]
That's because I talked about some "tweaking". Maybe some "hacking" is a better definition. I know that it will not work out of the box. But I suspect that CUDA can be upgraded using those packages with some "hacking" tecnique. I want to try something,but before I need to collect some relevant informations.

**Comment by ichergui** [comment]
Hi @Marietto2008 
Could you please share your motivation about upgrading the Linux kernel and CUDA for Jetson Nano ? 
This is not an easy task. You need to think about the whole SW stack not only Linux kernel and the compute stack.
FYI, Jetson Nano series are not support in JetPack 5 and JetPack 6.

**Comment by Marietto2008** [comment]
The Jetson nano is still a good soc for small / medium computing tasks. The Jetson Orin is too much expensive. A lot of users can't spend so much money for the Orin when for them the Nano is enough. 
I think that it should not die. Indipendent developers and system admins can fill the gap by refreshing the SW stack,because nVidia does not really care for their users. If they care,they would have created an intermediate model that would have cost not more than 200 EU. I've just upgraded Ubuntu 18.04 to Ubuntu 22.04 on the Jetson nano,so I have some hope to be able to upgrade also the CUDA libraries.

**Reply by ichergui** [reply]
Thanks @Marietto2008 for your feedback. 
I will think about your needs over the weekend and will get back to you

**Reply by harleylara** [reply]
I want to endorse @Marietto2008's position on this, we share a similar point of view... maybe this post will expand a bit on this [Let's Talk about Jetson Nano in 2024](https://wiki.eolab.de/doku.php?id=blog:lets-talk-about-jetson-nano-in-2024)
